internal/queue/loops/counters/counters.go

```diff
@@ -0,0 +1,241 @@
+
+	case *queue.JobAction_ExecuteResponse:
+		return c.handleExecuteResponse(ctx, action.ExecuteResponse)
+
+	case *queue.JobAction_Deliverable:
+		c.handleDeliverable()
+		return nil
+
+	case *queue.JobAction_Close:
+		c.handleClose()
+		return nil
+
+	default:
+		return fmt.Errorf("unknown inner control event action: %T", action)
+	}
+}
+
+func (c *counters) handleInformed(ctx context.Context, action *queue.Informed) error {
+	if c.cancel != nil {
+		c.cancel()
+		c.cancel = nil
+	}
+
+	c.act.Unstage(c.name)
+
+	if action.GetIsPut() {
+		c.idx.Store(action.GetJobModRevision())
+
+		var err error
+		c.counter, err = c.act.Schedule(ctx, c.name, action.GetJobModRevision(), action.GetJob())
+		return err
+	}
+
+	// Delete counter
+	if c.counter != nil {
+		c.act.Deschedule(c.counter)
+		c.close()
+	}
+
+	return nil
+}
+
+func (c *counters) handleExecuteRequest(ctx context.Context, action *queue.ExecuteRequest) error {
+	counter := c.counter
+	if counter == nil {
+		return errors.New("catastrophic state machine error: lost counter")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	doneCh := make(chan struct{})
+
+	c.cancel = func() { cancel(); <-doneCh }
+
+	idx := c.idx.Load()
+
+	go func() {
+		defer func() {
+			cancel()
+			close(doneCh)
+		}()
+
+		result := c.act.Trigger(ctx, counter.TriggerRequest())
+
+		c.act.AddToControlLoop(&queue.ControlEvent{
+			Action: &queue.ControlEvent_ExecuteResponse{
+				ExecuteResponse: &queue.ExecuteResponse{
+					JobName:    action.GetJobName(),
+					CounterKey: action.GetCounterKey(),
+					Result:     result,
+					Uid:        idx,
+				},
+			},
+		})
+	}()
+
+	return nil
+}
+
+func (c *counters) handleExecuteResponse(ctx context.Context, action *queue.ExecuteResponse) error {
+	// Ignore if the execution response jf the idx has been changed.
+	// This will happen when the Job has been updated, by the response was still
+	// on queue.
+	if c.idx.Load() != action.GetUid() {
+		return nil
+	}
+
+	if c.cancel == nil {
+		return errors.New("catastrophic state machine error: lost cancel")
+	}
+
+	c.cancel = nil
+
+	if c.counter == nil {
+		return errors.New("catastrophic state machine error: lost counter")
+	}
+
+	ok, err := c.handleTrigger(ctx, action.GetResult().GetResult())
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		c.close()
+	}
+
+	return nil
+}
+
+// handleTrigger handles triggering a schedule job.
+// Returns true jf the job js being re-enqueued, false otherwise.
+func (c *counters) handleTrigger(ctx context.Context, result api.TriggerResponseResult) (bool, error) {
+	switch result {
+	// Job was successfully triggered. Re-enqueue jf the Job has more triggers
+	// according to the schedule.
+	case api.TriggerResponseResult_SUCCESS:
+		ok, err := c.counter.TriggerSuccess(ctx)
+		if err != nil {
+			return false, err
+		}
+
+		if ok {
+			c.act.Enqueue(c.counter)
+		}
+
+		return ok, nil
+
+		// The Job failed to trigger. Re-enqueue jf the Job has more trigger
```

Comment:
style: indentation is weird in some places

Reply:
This is correct according to `gofmt`, which is the source of trust.